feat(AIP-121): adding guidance around consistency (#1060)
googleapis/aip: aip/general/0121.md

traditional RPC APIs, which can be used to implement common programming
patterns, such as database transactions, import and export, or data analysis.

> ### Consistency

Reviewer: "Read-after-write consistency"

Otherwise "consistency" is a very overloaded term.

Reviewer: Or maybe "strong consistency".

Do we care about write-after-write consistency also?

Author: Definition-wise I was hoping to leverage the term from the CAP theorem: https://en.wikipedia.org/wiki/CAP_theorem: "Every read receives the most recent write or an error." Although I imagine we have to qualify this a bit better.

Or I could title the section "Consistency Guarantees", since that generally implies distributed systems?

write-after-write... I didn't think it really mattered, since one can always validate whether their intended state has been applied and re-read again. 

For those who do want to have guarantee around their resource state's write, I figured AIP-154's etag might be helpful?

For methods that operate on a resource, the completion of those operations

> (either successful or with an error) **must** mean that the state of the

Reviewer: Do we need to mention LRO here, or is that understood?

Author: I feel like it's understood, since "long-running operation" is a variation of "operation", after all.

But I see the value in clarifying that it's both.

resource's existence and all user-settable values have reached a steady-state.

[output only][] values **should** also have reached a steady-state, except
for values that are related to the resource [state][].

Examples include:

- A get request for a resource immediately after the successful completion of a
  create method **must** return the resource.
- A get request for a resource immediately following an update **must** return

>   a resource with the final values from the update request.

Reviewer: What about delete?

Author: ack. These were just examples, but I suppose it doesn't hurt to add a delete.

traditional RPC APIs, which can be used to implement common programming
patterns, such as database transactions, import and export, or data analysis.

### Strong Consistency

> For methods that operate on the [managment plane][], the completion of those

Reviewer 2: ```suggestion
For methods that operate on the [management plane][], the completion of those
```

Author: applied.

operations (either successful or with an error, LRO or synchronous) **must**
mean that the state of the resource's existence and all user-settable values
have reached a steady-state.

> [output only][] values **should** also have reached a steady-state, except

Reviewer 2: What does this mean?

Author: basically that output-only values should be stable if they aren't related to resource state. E.g. they're not jittering around in values. This is for the same reason the operation itself should be consistent: the user should expect resource creation marks some event where the resource is more or less stable, since declarative clients have trouble reconciling transient state, and users will be surprised if the resource is constantly changing.

Of course being related to resource state is a little vague, since the only definition of state we have is AIP-216.

Reviewer 2: Like server generated IDs? create_time ?

Author: those are good examples of things that are steady state. AIP-216's state field is an example of what shouldn't be.

Would adding examples help? Or do you think this statement is too ambiguous?

> [output only][] values **should** also have reached a steady-state, except

Reviewer 2: ```suggestion
[output only][] values unrelated to the resource [state][] **should** also have reached a steady-state.
```

Author: applied.

for values that are related to the resource [state][].

Examples include:

> - Following a successful create, a get request for a resource **must**

Reviewer 2: Unless there's a race condition whereby it got deleted immediately. Same below for update.

If two people are updating, (unless the service handles stale state), then last one will win.

Author: good call! I updated the working to state that it has to be the latest mutation on the resource (should clarify that last operation wins). WDYT?

  return the resource.
- Following a successful update, a get request for a resource **must**

>   return the finale values from the update request.

Reviewer 2: ```suggestion
  return the final values from the update request.
```

Author: applied.
